mapie/calibration.py

```diff
@@ -0,0 +1,502 @@
+            Note that the sample weight defined are only for the training, not
+            for the calibration procedure.
+            By default ``None``.
+        estimator : ClassifierMixin
+            Estimator fitted.
+        calibrator : RegressorMixin
+            Calibrator to train.
+
+        Returns
+        -------
+        Dict[int, RegressorMixin]
+            Dictionnary of fitted calibrators.
+        """
+        X, y = indexable(X, y)
+        y = _check_y(y)
+        y_pred_calib = estimator.predict_proba(X=X)
+        max_prob, max_prob_arg = self._get_labels(
+            cast(str, self.method),
+            y_pred_calib
+        )
+        calibrators = {}
+        for item in np.unique(max_prob_arg):
+            calibrator_ = self._fit_calibrator(
+                item,
+                calibrator,
+                y,
+                max_prob,
+                max_prob_arg,
+                sample_weight,
+            )
+            calibrators[item] = calibrator_
+        return calibrators
+
+    def _pred_proba_calib(
+        self,
+        idx: int,
+        item: int,
+        calibrated_values: NDArray,
+        max_prob: NDArray,
+        max_prob_arg: NDArray,
+        calibrators: Dict[int, RegressorMixin],
+    ) -> NDArray:
+        """
+        Using the predicted probabilities, we calibrate new values with
+        the fitted calibrators. Note that if there is no calibrator for a
+        the specific class, then we simply output the not calibrated values.
+
+        Note that if the calibrated probability prediction is 0, there would
+        be an issue when finding the class is belongs to. We set it equal to
+        0.00001 as this would likely not have large impact on the calibration
+        scores, yet, would set the maximum to the correct label.
+
+
+        Parameters
+        ----------
+        item : int
+            The class value to be calibrated.
+        calibrated_values : NDArray
+            Array of calibrated values.
+        max_prob : NDArray of shape (n_samples,)
+            Values to be calibrated.
+        max_prob_arg : NDArray of shape (n_samples,)
+            Indicator of the values to be calibrated.
+        calibrators : Dict[RegressorMixin] of len n_classes in calibration set
+            Dictionary of all the previously fitted calibrators.
+
+        Returns
+        -------
+        NDArray
+            Updated calibrated values from the predictions of the calibrators.
+
+        Warnings
+            If there has not been a fitted calibrator for the specific class.
+        """
+        correct_label = np.where(max_prob_arg.ravel() == item)[0].ravel()
+        if item not in calibrators:
+            calibrated_values[
+                correct_label, idx
+                ] = max_prob[correct_label].ravel()
+            warnings.warn(
+                "WARNING: This calibration was not previously seen"
```

Comment:
This predicted label has not been seen during the calibration and ....

Reply:
done.